python/lsst/daf/butler/formatters/jsonFormatter.py

```diff
@@ -78,7 +78,50 @@ def _writeFile(self, inMemoryDataset, fileDescriptor):
         with open(fileDescriptor.location.path, "w") as fd:
             if hasattr(inMemoryDataset, "_asdict"):
                 inMemoryDataset = inMemoryDataset._asdict()
-            json.dump(inMemoryDataset, fd)
+            fd.write(self._toBytes(inMemoryDataset).decode())
+
+    def _fromBytes(self, bytesObject, pytype=None):
+        """Read the bytes object as a python object.
+
+        Parameters
+        ----------
+        serializedDataset : `bytes`
+            Bytes object to unserialize.
+        pytype : `class`, optional
```

Comment:
Do we ever use it? I think the coerceType code elsewhere works around the problem that JSON does not return python objects.

Reply:
I just followed what other classes (specifically their `_readFile` and `_writeFile` methods) were doing; and they explicitly mentioned `pytype` even if the docstring was `not used by this implementation`. Should I change this (just here or for all of them)?

Reply:
If it's used everywhere else it should stay here. I can always take another look later to see if things can be tidied up if nothing ever uses it.

Reply:
To save you some trouble (I guess), it's `FitsCatalogFormatter` and `PexConfigFormatter` that use it. I assumed that they were there to match the declaration from `FileFormatter`

[`FileFormatter`](https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/formatters/fileFormatter.py)
```
class FileFormatter(Formatter):
    @abstractmethod
    def _readFile(self, path, pytype=None):
...
        pytype : `class`, optional
            Class to use to read the file.
...
```

since they were left in (on master) for the rest of the formatters:

[`JsonFormatter`](https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/formatters/jsonFormatter.py#L38)
```
class JsonFormatter(FileFormatter):
    def _readFile(self, path, pytype=None):
```
(pytype unused, kwarg not documented)

[`PickleFormatter`](https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/formatters/pickleFormatter.py#L40)
```   
class PickleFormatter(FileFormatter):
...
     pytype : `class`, optional
            Not used by this implementation.
...
```
(pytype unused, kwarg documented)

[`YamlFormatter`](https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/formatters/yamlFormatter.py#L38)
``` 
class YamlFormatter(FileFormatter):
    def _readFile(self, path, pytype=None):
```
(pytype unused, kwarg not documented)

and the exception to the rule is the [`FitsExposureFormatter`](https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/formatters/fitsExposureFormatter.py) but that class seems special. On retrospective it feels like perhaps I should not have left them in the definitions. What I have done is left the doc strings for `_readFile` and `_writeFile` on the formatters intact and then I effectively copied their declaration for `_readBytes` and `_writeBytes`. This now looks like (`JsonFormatter` for example):

[`JsonFormatter`](https://github.com/DinoBektesevic/daf_butler/blob/u/DinoBektesevic/merged_s3rds/python/lsst/daf/butler/formatters/jsonFormatter.py)
```
class JsonFormatter(FileFormatter):
...
    def _readFile(self, path, pytype=None):
(pytype not used, kwarg not documented)
```
```
...
    def _fromBytes(self, serializedDataset, pytype=None):
...
        pytype : `class`, optional
            Not used by this implementation.
...
```
(pytype not used, kwarg documented)

Which now feels weird and clunky. It'd probably be best to remove the `pytype` argument, I don't think `abstractmethod` would raise an error, but it could be weird because you'd usually expect to see it there. Any thoughts on the topic are welcomed and will be implemented.